### PR TITLE
MC-137: Add product associations export in product export code

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | y
| Specs pass?*         | y
| Checkstyle issues?** | n
| PMD issues?***       | n
| Fixed tickets        | MC-137
| Doc PR               |

### CODE REVIEW

File: Normalizer/AssociationNormalizer.php [hunk 1]
```diff
@@ -35,26 +35,16 @@ public function __construct(ValidProductHelper $validProductHelper)
     /**
      * {@inheritdoc}
      */
-    public function normalize($object, $format = null, array $context = [])
+    public function normalize($association, $format = null, array $context = [])
```
Comment (fitn) [inline Normalizer/AssociationNormalizer.php]: No $object is better. It respects NormalizerInterface and it's homogeneous with your code.

File: Processor/ProductToArrayProcessor.php [hunk 2]
```diff
@@ -55,21 +65,51 @@ public function process($item)
...
+                'status'            => 'enabled'
             ]
         ];
 
-        return $this->normalizer->normalize($item, 'api_import', $context);
+        $product = $this->normalizer->normalize($item, 'api_import', $context);
+
+        foreach ($item->getAssociations() as $association) {
+            $associationPart = $this->normalizer->normalize($association, 'api_import', $context);
+            if (is_array($associationPart)) {
```
Comment (fitn) [inline Processor/ProductToArrayProcessor.php]: Why don't you return an array everytime? empty or not..

File: Processor/ProductToArrayProcessor.php [hunk 3]
```diff
@@ -55,21 +65,49 @@ public function process($item)
...
+                'X_SELL'  => 'crosssell',
+                'RELATED' => 'related',
+                'PACK'    => ''
+            ],
+            'attributeMapping'    => [
+                'sku'               => 'sku',
+                'name'              => 'name',
+                'description'       => 'description',
+                'short_description' => 'short_description',
+                'status'            => 'enabled'
```
Comment (fitn) [inline Processor/ProductToArrayProcessor.php]: Don't forget to add tasks for these improvements. Association and Attribute mappings should not be hardcoded ;)

File: Processor/ProductToArrayProcessor.php [hunk 4]
```diff
@@ -55,21 +65,49 @@ public function process($item)
...
             ]
         ];
 
-        return $this->normalizer->normalize($item, 'api_import', $context);
+        $product = $this->normalizer->normalize($item, 'api_import', $context);
+
+        foreach ($item->getAssociations() as $association) {
+            $associationPart = $this->normalizer->normalize($association, 'api_import', $context);
+            $product = array_merge($product, $associationPart);
+        }
```
Comment (fitn) [inline Processor/ProductToArrayProcessor.php]: I think Product normalization should call Association one.

File: Normalizer/ProductNormalizer.php [hunk 5]
```diff
@@ -53,6 +53,11 @@ public function normalize($object, $format = null, array $context = [])
             ];
         }
 
+        foreach ($object->getAssociations() as $association) {
+            $associationPart = $this->normalizer->normalize($association, 'api_import', $context);
+            $processedProduct = array_merge($processedProduct, $associationPart);
```
Comment (fitn) [inline Normalizer/ProductNormalizer.php]: $assocationPart = $this->normalizeAssociations($association, $format, $context);
$processedProduct = array_merge($processedProduct, $associationPart);

protected function normalizeAssociations($object, ...)
{
    return $this->normalizer->normalize($object->getAssociations(), $format, $context);
}

File: spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php [hunk 6]
```diff
@@ -106,7 +110,22 @@ public function it_normalizes_a_product_to_an_array(
...
+            ]
+        ]);
+        $normalizer->normalize($associationUpSell, 'api_import', $context)->shouldBeCalled()->willReturn([
+            [
+                '_links_upsell_sku' => 'sku-003'
+            ],
+            [
+                '_links_upsell_sku' => 'sku-002'
+            ]
+        ]);
```
Comment (fitn) [inline spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php]: Add spec without associations

File: Normalizer/ProductNormalizer.php [hunk 7]
```diff
@@ -53,7 +53,9 @@ public function normalize($object, $format = null, array $context = [])
             ];
         }
 
-        return $processedProduct;
+        $associationParts = $this->normalizer->normalize($object->getAssociations(), 'api_import', $context);
```
Comment (fitn) [inline Normalizer/ProductNormalizer.php]: Better in a submethod

File: Normalizer/ProductNormalizer.php [hunk 8]
```diff
@@ -53,7 +53,9 @@ public function normalize($object, $format = null, array $context = [])
...
+
+        return is_array($associationParts) ? array_merge($processedProduct, $associationParts[0]) : $processedProduct;
```
Comment (fitn) [inline Normalizer/ProductNormalizer.php]: I don't understand this step. is_array means you can have an empty array and then you merge only the first part of your array..

File: Normalizer/AssociationNormalizer.php (hunk 1 above)
Comment (BitOne) [inline Normalizer/AssociationNormalizer.php]: Nop, association is clearer for the code. The NormalizerInterface is still respected.

File: Normalizer/AssociationNormalizer.php [hunk 9]
```diff
@@ -35,33 +33,23 @@ public function __construct(ValidProductHelper $validProductHelper)
...
     {
         $associations       = [];
         $channel            = $context['channel'];
-        $validProducts      = $this->validProductHelper->getValidProducts($channel, $object->getProducts());
+        $validProducts      = $this->validProductHelper->getValidProducts($channel, $association->getProducts());
         $associationMapping = $context['associationMapping'];
-        $typeCode           = $associationMapping[$object->getAssociationType()->getCode()];
+        $typeCode           = $associationMapping[$association->getAssociationType()->getCode()];
```
Comment (BitOne) [inline Normalizer/AssociationNormalizer.php]: assocTypeCode

File: Normalizer/AssociationNormalizer.php [hunk 10]
```diff
@@ -35,33 +33,23 @@ public function __construct(ValidProductHelper $validProductHelper)
     /**
      * {@inheritdoc}
      */
-    public function normalize($object, $format = null, array $context = [])
+    public function normalize($association, $format = null, array $context = [])
     {
         $associations       = [];
         $channel            = $context['channel'];
-        $validProducts      = $this->validProductHelper->getValidProducts($channel, $object->getProducts());
+        $validProducts      = $this->validProductHelper->getValidProducts($channel, $association->getProducts());
```
Comment (BitOne) [inline Normalizer/AssociationNormalizer.php]: What are validProducts ? Why they could be not valid ?

Comment (willy-ahva) [inline Normalizer/AssociationNormalizer.php]: with getAssociations you get all associated products. This method help to retrieve valids ones, ie.   those which are in the channel and complete. So yes, you can have "invalid" products.

File: Normalizer/ProductNormalizer.php (hunk 8 above)
Comment (willy-ahva) [inline Normalizer/ProductNormalizer.php]: if there is something to normalize (ie getAssociations doesn't return an empty collection) normalizer returns an array else it returns null. I can't change tis behavior without changing lot of behavior for others normalizers. Return for this type of data is always like [ [ [ my data] ] ]. So I need to keep the first index to merge.

File: spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php [hunk 11]
```diff
@@ -106,6 +110,136 @@ public function it_normalizes_a_product_to_an_array(
...
+                [
+                    '_links_crosssell_sku' => 'sku-001'
+                ],
+                [
+                    '_links_crosssell_sku' => 'sku-002'
+                ]
+            ]
+        ]);
+
+        $this->setSerializer($normalizer);
```
Comment (fitn) [inline spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php]: Should be in let method

Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php]: Nop, only beConstructedWith() should be in the let. It allows to have method totally independent.

Comment (fitn) [inline spec/Pim/Bundle/MagentoConnectorBundle/Normalizer/ProductNormalizerSpec.php]: You can't have a normalizer without calling the setSerializer, it's part of the object instanciation.

File: Normalizer/ProductNormalizer.php [hunk 12]
```diff
@@ -156,4 +158,19 @@ protected function getProductCategories(ProductInterface $product, $format, arra
...
+     * @param ProductInterface $product
+     * @param array            $context
+     *
+     * @return array
+     */
+    protected function getNormalizedAssociation(ProductInterface $product, array $context)
+    {
+        $associationParts = $this->normalizer->normalize($product->getAssociations(), 'api_import', $context);
+
+        return is_array($associationParts) ? $associationParts[0] : [];
```
Comment (fitn) [inline Normalizer/ProductNormalizer.php]: Same than before... What about empty array? is_array(array()) returns true, then $associationParts will returns an undefined index 0..
Why did you only get the first index? You want only the "first" association here? MySQL get the associations in a random order.

**Comment by fitn** [comment]
GTM

**Pull request merged**